File: 1579-remove-max-number-of-edges-to-keep-graph-fully-traversable/1579-remove-max-number-of-edges-to-keep-graph-fully-traversable.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {number}
 */
var maxNumEdgesToRemove = function(n, edges) {
  edges.sort((a, b) => b[0] - a[0]);
  const ufAlice = new UnionFind(n + 1);
  const ufBob = new UnionFind(n + 1);
  let count = 0;
  
  for (let [type, u, v] of edges) {
    if (type === 3) count += ufAlice.union(u, v) | ufBob.union(u, v);
    else if (type === 2) count += ufBob.union(u, v);
    else count += ufAlice.union(u, v);
  }
  
  return ufAlice.united() && ufBob.united() ? edges.length - count : -1;
};

class UnionFind {
  constructor(n) {
    this.parent = Array.from({ length: n }, (_, i) => i);
    this.count = n;
  }
  
  find(x) {
    let parent = this.parent;
    while (x !== parent[x]) {
      parent[x] = parent[parent[x]];
      x = parent[x];
    }
    
    return x;
  }
  
  united() {
    return this.count === 2;
  }
  
  union(x, y) {
    let rootX = this.find(x);
    let rootY = this.find(y);
    if (rootX === rootY) return false;
    
    this.parent[rootX] = rootY;
    this.count--;
    
    return true;
  }
}

File: 1579-remove-max-number-of-edges-to-keep-graph-fully-traversable/1579-remove-max-number-of-edges-to-keep-graph-fully-traversable.py
class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
        self.count = n
        
    def find(self, x):
        if x != self.parent[x]:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    
    def united(self):
        return self.count == 2
    
    def union(self, x, y):
        root_x, root_y = self.find(x), self.find(y)
        if root_x == root_y:
            return False
        self.parent[root_x] = root_y
        self.count -= 1
        return True
      

class Solution:
    def maxNumEdgesToRemove(self, n: int, edges: List[List[int]]) -> int:
        ufa = UnionFind(n + 1)
        ufb = UnionFind(n + 1)
        count = 0
        for t, u, v in sorted(edges, reverse=True):
            if t == 3:
                count += ufa.union(u, v) | ufb.union(u, v)
            elif t == 2:
                count += ufb.union(u, v)
            else:
                count += ufa.union(u, v)
        return len(edges) - count if ufa.united() and ufb.united() else -1